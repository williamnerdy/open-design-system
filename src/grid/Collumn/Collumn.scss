@import '../../styles/variables';

.col {
  display: block;

  @for $i from 1 through $grid-collumns {
    &.size-#{$i} {
      width: percentage($i / $grid-collumns);
    }
  }

  &.align-left {
    align-self: flex-start;
  }

  &.align-center {
    align-self: center;
  }

  &.align-right {
    align-self: end;
  }
}
