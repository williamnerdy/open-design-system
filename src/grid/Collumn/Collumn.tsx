import React from 'react';
import classNames from 'classnames';

type GridSizes = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;

type Aligns = 'left' | 'center' | 'right';

type Mixed = { size?: GridSizes; align?: Aligns };

type Responsives = Mixed | GridSizes | Aligns;

interface Props {
  /** Children */
  children: React.ReactNode;
  /** Extreme small size */
  size?: GridSizes;
  /** Align content */
  align?: Aligns;
  /** Small size */
  sm?: Responsives;
  /** Medium size */
  md?: Responsives;
  /** Large size */
  lg?: Responsives;
}

const getResponsive = (name: string, val?: Responsives) => {
  const align = typeof val === 'object' && val.align ? val.align : val;
  const size = typeof val === 'object' && val.size ? val.size : val;

  return {
    [`${name}-${align}`]: align && typeof align !== 'object',
    [`${name}-${size}`]: size && typeof size !== 'object'
  };
};

const Collumn = ({ children, size, align, sm, md, lg }: Props) => {
  const classes = classNames('col', {
    [`size-${size}`]: size,
    [`align-${align}`]: align,
    ...getResponsive('sm', sm),
    ...getResponsive('md', md),
    ...getResponsive('lg', lg)
  });

  return <div className={classes}>{children}</div>;
};

export default Collumn;
