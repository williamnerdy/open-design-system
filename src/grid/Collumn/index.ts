import Collumn from './Collumn';
export default Collumn;
