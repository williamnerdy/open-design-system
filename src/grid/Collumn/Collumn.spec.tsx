import React from 'react';
import { cleanup, render } from '@testing-library/react';
import Collumn from './Collumn';

describe('<Collumn />', () => {
  const contentCollumn = 'Collumn content';
  beforeEach(cleanup);

  it('should render without prop', () => {
    const { asFragment, getByText } = render(
      <Collumn>{contentCollumn}</Collumn>
    );
    expect(asFragment()).toMatchSnapshot();
    expect(getByText(contentCollumn).classList.contains('col')).toBe(true);
  });

  it('should render with 1 size', () => {
    const { asFragment, getByText } = render(
      <Collumn size={1}>{contentCollumn}</Collumn>
    );
    expect(asFragment()).toMatchSnapshot();
    expect(getByText(contentCollumn).classList.contains('size-1')).toBe(true);
  });

  it('should render with 6 size', () => {
    const { asFragment, getByText } = render(
      <Collumn size={6}>{contentCollumn}</Collumn>
    );
    expect(asFragment()).toMatchSnapshot();
    expect(getByText(contentCollumn).classList.contains('size-6')).toBe(true);
  });

  it('should render with 12 size', () => {
    const { asFragment, getByText } = render(
      <Collumn size={12}>{contentCollumn}</Collumn>
    );
    expect(asFragment()).toMatchSnapshot();
    expect(getByText(contentCollumn).classList.contains('size-12')).toBe(true);
  });

  describe('Responsive sizes', () => {
    it('should render with sm 1 size', () => {
      const { asFragment, getByText } = render(
        <Collumn sm={1}>{contentCollumn}</Collumn>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentCollumn).classList.contains('sm-1')).toBe(true);
    });

    it('should render with md 1 size', () => {
      const { asFragment, getByText } = render(
        <Collumn md={1}>{contentCollumn}</Collumn>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentCollumn).classList.contains('md-1')).toBe(true);
    });

    it('should render with lg 1 size', () => {
      const { asFragment, getByText } = render(
        <Collumn lg={1}>{contentCollumn}</Collumn>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentCollumn).classList.contains('lg-1')).toBe(true);
    });
  });

  describe('align', () => {
    it('should render with align left', () => {
      const { asFragment, getByText } = render(
        <Collumn align="left">{contentCollumn}</Collumn>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentCollumn).classList.contains('align-left')).toBe(
        true
      );
    });

    it('should render with align center', () => {
      const { asFragment, getByText } = render(
        <Collumn align="center">{contentCollumn}</Collumn>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentCollumn).classList.contains('align-center')).toBe(
        true
      );
    });

    it('should render with align right', () => {
      const { asFragment, getByText } = render(
        <Collumn align="right">{contentCollumn}</Collumn>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentCollumn).classList.contains('align-right')).toBe(
        true
      );
    });
  });

  describe('align by size', () => {
    it('should render with align center sm', () => {
      const { asFragment, getByText } = render(
        <Collumn sm="center">{contentCollumn}</Collumn>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentCollumn).classList.contains('sm-center')).toBe(
        true
      );
    });

    it('should render with align center sm', () => {
      const { asFragment, getByText } = render(
        <Collumn sm={{ align: 'center' }}>{contentCollumn}</Collumn>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentCollumn).classList.contains('sm-center')).toBe(
        true
      );
    });

    it('should render with align center md', () => {
      const { asFragment, getByText } = render(
        <Collumn md="center">{contentCollumn}</Collumn>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentCollumn).classList.contains('md-center')).toBe(
        true
      );
    });

    it('should render with align center md', () => {
      const { asFragment, getByText } = render(
        <Collumn md={{ align: 'center' }}>{contentCollumn}</Collumn>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentCollumn).classList.contains('md-center')).toBe(
        true
      );
    });

    it('should render with align center lg', () => {
      const { asFragment, getByText } = render(
        <Collumn lg="center">{contentCollumn}</Collumn>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentCollumn).classList.contains('lg-center')).toBe(
        true
      );
    });

    it('should render with align center lg', () => {
      const { asFragment, getByText } = render(
        <Collumn lg={{ align: 'center' }}>{contentCollumn}</Collumn>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentCollumn).classList.contains('lg-center')).toBe(
        true
      );
    });
  });

  describe('align and size', () => {
    it('should render with align center and size 6 sm', () => {
      const { asFragment, getByText } = render(
        <Collumn sm={{ size: 6, align: 'center' }}>{contentCollumn}</Collumn>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentCollumn).classList.contains('sm-6')).toBe(true);
      expect(getByText(contentCollumn).classList.contains('sm-center')).toBe(
        true
      );
    });

    it('should render with align center and size 6 md', () => {
      const { asFragment, getByText } = render(
        <Collumn md={{ size: 6, align: 'center' }}>{contentCollumn}</Collumn>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentCollumn).classList.contains('md-6')).toBe(true);
      expect(getByText(contentCollumn).classList.contains('md-center')).toBe(
        true
      );
    });

    it('should render with align center and size 6 lg', () => {
      const { asFragment, getByText } = render(
        <Collumn lg={{ size: 6, align: 'center' }}>{contentCollumn}</Collumn>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentCollumn).classList.contains('lg-6')).toBe(true);
      expect(getByText(contentCollumn).classList.contains('lg-center')).toBe(
        true
      );
    });
  });
});
