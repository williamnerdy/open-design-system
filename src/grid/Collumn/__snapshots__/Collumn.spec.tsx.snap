// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Collumn /> Responsive sizes should render with lg 1 size 1`] = `
<DocumentFragment>
  <div
    class="col lg-1"
  >
    Collumn content
  </div>
</DocumentFragment>
`;

exports[`<Collumn /> Responsive sizes should render with md 1 size 1`] = `
<DocumentFragment>
  <div
    class="col md-1"
  >
    Collumn content
  </div>
</DocumentFragment>
`;

exports[`<Collumn /> Responsive sizes should render with sm 1 size 1`] = `
<DocumentFragment>
  <div
    class="col sm-1"
  >
    Collumn content
  </div>
</DocumentFragment>
`;

exports[`<Collumn /> align and size should render with align center and size 6 lg 1`] = `
<DocumentFragment>
  <div
    class="col lg-center lg-6"
  >
    Collumn content
  </div>
</DocumentFragment>
`;

exports[`<Collumn /> align and size should render with align center and size 6 md 1`] = `
<DocumentFragment>
  <div
    class="col md-center md-6"
  >
    Collumn content
  </div>
</DocumentFragment>
`;

exports[`<Collumn /> align and size should render with align center and size 6 sm 1`] = `
<DocumentFragment>
  <div
    class="col sm-center sm-6"
  >
    Collumn content
  </div>
</DocumentFragment>
`;

exports[`<Collumn /> align by size should render with align center lg 1`] = `
<DocumentFragment>
  <div
    class="col lg-center"
  >
    Collumn content
  </div>
</DocumentFragment>
`;

exports[`<Collumn /> align by size should render with align center lg 2`] = `
<DocumentFragment>
  <div
    class="col lg-center"
  >
    Collumn content
  </div>
</DocumentFragment>
`;

exports[`<Collumn /> align by size should render with align center md 1`] = `
<DocumentFragment>
  <div
    class="col md-center"
  >
    Collumn content
  </div>
</DocumentFragment>
`;

exports[`<Collumn /> align by size should render with align center md 2`] = `
<DocumentFragment>
  <div
    class="col md-center"
  >
    Collumn content
  </div>
</DocumentFragment>
`;

exports[`<Collumn /> align by size should render with align center sm 1`] = `
<DocumentFragment>
  <div
    class="col sm-center"
  >
    Collumn content
  </div>
</DocumentFragment>
`;

exports[`<Collumn /> align by size should render with align center sm 2`] = `
<DocumentFragment>
  <div
    class="col sm-center"
  >
    Collumn content
  </div>
</DocumentFragment>
`;

exports[`<Collumn /> align should render with align center 1`] = `
<DocumentFragment>
  <div
    class="col align-center"
  >
    Collumn content
  </div>
</DocumentFragment>
`;

exports[`<Collumn /> align should render with align left 1`] = `
<DocumentFragment>
  <div
    class="col align-left"
  >
    Collumn content
  </div>
</DocumentFragment>
`;

exports[`<Collumn /> align should render with align right 1`] = `
<DocumentFragment>
  <div
    class="col align-right"
  >
    Collumn content
  </div>
</DocumentFragment>
`;

exports[`<Collumn /> should render with 1 size 1`] = `
<DocumentFragment>
  <div
    class="col size-1"
  >
    Collumn content
  </div>
</DocumentFragment>
`;

exports[`<Collumn /> should render with 6 size 1`] = `
<DocumentFragment>
  <div
    class="col size-6"
  >
    Collumn content
  </div>
</DocumentFragment>
`;

exports[`<Collumn /> should render with 12 size 1`] = `
<DocumentFragment>
  <div
    class="col size-12"
  >
    Collumn content
  </div>
</DocumentFragment>
`;

exports[`<Collumn /> should render without prop 1`] = `
<DocumentFragment>
  <div
    class="col"
  >
    Collumn content
  </div>
</DocumentFragment>
`;
